### PR TITLE
Optimizied the fetch_all_features endpoint

### PR DESCRIPTION
**Pull request by dusktreader**
#### What
Optimized the fetch_all_features endpoint to eliminate extra queries.

#### Why
This endpoint, among others, was spawning too many cascading queries to load data.

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

### CODE REVIEW

File: agent/lm_agent/backend_utils/models.py [hunk 1]
```diff
@@ -41,6 +41,22 @@ class FeatureSchema(BaseModel):
     booked_total: int
 
 
+class FlatFeatureSchema(BaseModel):
```
Comment (fschuch) [inline agent/lm_agent/backend_utils/models.py]: I'm concerned about this being incompatible with the current implementation of the UI, I see CLI was changed on the PR as well. Why not keep the previous schema?

File: backend/lm_backend/api/models/feature.py [hunk 2]
```diff
@@ -39,13 +39,12 @@ def booked_total(self) -> int:
         """
         return sum(b.quantity for b in self.bookings)
 
-    @booked_total.inplace.expression
-    @classmethod
-    def _booked_total_expression(cls) -> SQLColumnExpression[int]:
+    @hybrid_property
+    def product_name(self) -> int:
```
Comment (julianaklulo) [inline backend/lm_backend/api/models/feature.py]: Why is this one needed when we have the `product_name` returned by the query?

File: lm-cli/lm_cli/subapps/features.py [hunk 3]
```diff
@@ -83,7 +86,7 @@ def list_all(
         ),
     )
 
-    formatted_data = format_data(feature_data)
+    formatted_data = format_data(feature_data, flat=True)
```
Comment (julianaklulo) [inline lm-cli/lm_cli/subapps/features.py]: If we're always running `format_data` with `flat=True`, why do we need the flag?

File: backend/lm_backend/api/cruds/feature.py [hunk 4]
```diff
@@ -88,3 +90,38 @@ async def filter_by_product_feature_and_client_id(
...
+        sort_field: Optional[str] = None,
+        sort_ascending: bool = True,
+    ) -> List[FlatFeatureSchema]:
+        """
+        Read all objects.
+        Returns a list of objects.
+        """
+        try:
+            stmt = (
+                select(
```
Comment (julianaklulo) [inline backend/lm_backend/api/cruds/feature.py:107]: Nice query!

**Review by julianaklulo — CHANGES_REQUESTED** [review]

File: lm-cli/lm_cli/subapps/features.py (hunk 3 above)
Comment (dusktreader) [reply lm-cli/lm_cli/subapps/features.py]: I wasn't using the `flat` flag for the `get_one` subcommand (line 125).

File: backend/lm_backend/api/models/feature.py (hunk 2 above)
Comment (dusktreader) [reply backend/lm_backend/api/models/feature.py]: The idea was to make it possible to use the `FlatFeatureSchema` with a single model instance as well as the data brought back by the `read_all()` method. I guess we aren't using it at the moment.

File: agent/lm_agent/backend_utils/models.py (hunk 1 above)
Comment (dusktreader) [reply agent/lm_agent/backend_utils/models.py]: I unflattened it. Instead, I build the objects for the nested schema from the flat results from the db query.

File: backend/lm_backend/api/models/feature.py (hunk 2 above)
Comment (dusktreader) [reply backend/lm_backend/api/models/feature.py]: Removed in commit 8eb6c76

File: agent/lm_agent/backend_utils/models.py (hunk 1 above)
Comment (dusktreader) [reply agent/lm_agent/backend_utils/models.py]: Resolved in commit 8eb6c76

File: lm-cli/lm_cli/subapps/features.py (hunk 3 above)
Comment (dusktreader) [reply lm-cli/lm_cli/subapps/features.py]: No longer relevant as I unflattened it in commit 8eb6c76

File: backend/lm_backend/api/routes/products.py [hunk 5]
```diff
@@ -11,7 +11,11 @@
 router = APIRouter()
 
 
-crud_product = GenericCRUD(Product, ProductCreateSchema, ProductUpdateSchema)
+class MagicCRUD(GenericCRUD):
```
Comment (julianaklulo) [inline backend/lm_backend/api/routes/products.py]: Why only `Product` needs `MagicCRUD`?

File: backend/lm_backend/api/cruds/generic.py [hunk 6]
```diff
@@ -174,3 +169,15 @@ async def delete(self, db_session: AsyncSession, id: Union[Column[int], int]):
             raise HTTPException(status_code=400, detail=f"{self.model.__name__} could not be deleted.")
 
         return {"message": f"{self.model.__name__} deleted successfully."}
+
+
+class GenericCRUD(_GenericCRUD):
```
Comment (julianaklulo) [inline backend/lm_backend/api/cruds/generic.py]: That's so weird, why is `mypy` complaining about this just now?

File: backend/lm_backend/api/routes/products.py (hunk 5 above)
Comment (dusktreader) [reply backend/lm_backend/api/routes/products.py]: This was an experiment. I forgot to put away the test tubes afterward 😓

File: backend/lm_backend/api/cruds/generic.py (hunk 6 above)
Comment (dusktreader) [reply backend/lm_backend/api/cruds/generic.py]: I have no idea why this suddenly started.

Comment (julianaklulo) [reply backend/lm_backend/api/cruds/generic.py]: I see you fixed it, nice!

File: backend/lm_backend/api/routes/jobs.py [hunk 7]
```diff
@@ -43,10 +42,13 @@ async def create_job(
     if job.cluster_client_id is None:
         job.cluster_client_id = client_id
 
-    job_created: Job = await crud_job.create(
+    job_created = await crud_job.create(
         db_session=secure_session.session, obj=JobCreateSchema(**job.dict(exclude={"bookings"}))
     )
 
+    # Appease static type checkers
```
Comment (julianaklulo) [inline backend/lm_backend/api/routes/jobs.py]: You removed it from elsewhere, is it still needed?

File: backend/lm_backend/api/routes/jobs.py [hunk 8]
```diff
@@ -166,12 +167,16 @@ async def delete_job_by_slurm_id(
...
 
-    jobs: List[Job] = await crud_job.filter(
+    jobs = await crud_job.filter(
         db_session=secure_session.session,
         filter_expressions=[Job.slurm_job_id == slurm_job_id, Job.cluster_client_id == client_id],
     )
 
     for job in jobs:
+
+        # Appease static type checkers
```
Comment (julianaklulo) [inline backend/lm_backend/api/routes/jobs.py]: Here as well.

**Review by julianaklulo — APPROVED** [review]
Just two small comments regarding mypy.
Overall it looks pretty good, nice work!

File: backend/lm_backend/api/routes/jobs.py (hunk 7 above)
Comment (dusktreader) [reply backend/lm_backend/api/routes/jobs.py]: Let me check...

File: backend/lm_backend/api/routes/jobs.py (hunk 8 above)
Comment (dusktreader) [reply backend/lm_backend/api/routes/jobs.py]: Removed in commit ed929c4

File: backend/lm_backend/api/routes/jobs.py (hunk 7 above)
Comment (dusktreader) [reply backend/lm_backend/api/routes/jobs.py]: Removed in commit ed929c4

**Pull request merged**